# Data Wrangling and Subsetting
### This script contains the following points:

#### 01. Importing libraries
#### 02. Importing orders.csv, products.csv and departments.csv data¶
#### 03. Checking orders.csv, products.csv and departments.csv dataframes
#### 04. Wrangling orders.csv data
* Value Counts for the orders dataframe columns
* Renaming order_dow column
* Changing the orders dataframe order_id column data type

#### 05. Transposing Data
#### 06. Data Dictionary
#### 07. Subsetting
#### 08. Querying Data

## 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing orders.csv, products.csv and departments.csv data

In [2]:
# Assign the main project folder path to the variable path and display the path
path = r'/Users/elsaekevall/Jupyter_Notebook/Career_Foundry/06_2022_Instacart_Basket_Analysis/'
path

'/Users/elsaekevall/Jupyter_Notebook/Career_Foundry/06_2022_Instacart_Basket_Analysis/'

In [3]:
# Use the os.path.join() function to import the orders.csv, products.csv and departments.csv file as pandas dataframes without the index column
df_ords = pd.read_csv(os.path.join(path, '02_Data', '02_1_Original_Data', 'orders.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02_Data', '02_1_Original_Data', 'products.csv'), index_col = False)
df_dep = pd.read_csv(os.path.join(path, '02_Data', '02_1_Original_Data', 'departments.csv'), index_col = False)

## 03. Checking orders.csv, products.csv and departments.csv dataframes

In [4]:
# Use the df.head function to view the first five rows of the orders.csv dataframe
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [5]:
# Use the df.head function to view the first five rows of the products.csv dataframe
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
# Use the df.head function to view the first five rows of the departments.csv dataframe
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

## 04.Wrangling orders.csv data

### Dropping eval_set column in orders dataframe

In [7]:
# Dropping eval_set column from orders.csv and overwriting the original dataframe then view using head() function
df_ords = df_ords.drop(columns = ['eval_set'])
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Value Counts for the orders dataframe columns

In [8]:
# View value counts for the order_id column
df_ords['order_id'].value_counts(dropna = False)

4098       1
440957     1
1735141    1
1731047    1
1710569    1
          ..
3244901    1
3248999    1
3269481    1
3273579    1
4094       1
Name: order_id, Length: 3421083, dtype: int64

In [9]:
# View value counts for the user_id column
df_ords['user_id'].value_counts(dropna = False)

82420     100
154678    100
134199    100
60040     100
143177    100
         ... 
109381      4
169577      4
167530      4
105283      4
185566      4
Name: user_id, Length: 206209, dtype: int64

In [10]:
# View value counts for the order_number column
df_ords['order_number'].value_counts(dropna = False)

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

In [11]:
# View value counts for the order_dow column
df_ords['order_dow'].value_counts(dropna = False)

0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: order_dow, dtype: int64

In [12]:
# View value counts for the order_hour_of_day column
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

In [13]:
# View value counts for the days_since_prior_order column
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

### Renaming order_dow column

In [14]:
# Rename the order_dow column to orders_day_of_week and view the first five rows of the renamed dataframe
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Changing the orders dataframe order_id column data type

In [15]:
# Dataframe information
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


In [16]:
# Change the order_id column and view the data types
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(1)
memory usage: 156.6+ MB


In [17]:
# Check data type of only the order_id column
df_ords['order_id'].dtype

dtype('O')

## 05. Transposing Data

In [18]:
# Use the T function to transpose the data and save the dataframe as df_dep_t then the .head() function to view 
df_dep_t = df_dep.T
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

I have removed the step below the line as I don't think it is needed - the output is exactly the same as above and without it you get the same results
***
*#reset the index and view the dataframe* <br>
df_dep_t.reset_index() <br>
df_dep_t

In [19]:
# use the iloc[] function (locate an integer) to set the new-header variable to 0 department
new_header = df_dep_t.iloc[0]
new_header

0    department
Name: department_id, dtype: object

In [20]:
# Create a new dataframe without the first row and view it
df_dep_t_new = df_dep_t[1:]
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [21]:
# Add new_header as the header row and view the df_dep_t_new dataframe
df_dep_t_new.columns = new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## 06. Data Dictionary

In [22]:
# Convert the df_dep_t_new dataframe to a data dictionary and view
data_dict_IC = df_dep_t_new.to_dict('index')
data_dict_IC

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [23]:
# How to look up department_id 19 using the data dictionary
print(data_dict_IC.get('19'))

{'department': 'snacks'}


## 07. Subsetting 

In [24]:
# Search for products in the same department and view the first five rows
df_snacks =  df_prods[df_prods['department_id']==19]
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [25]:
# Search for products in the same department using the loc function (locate a column) and view the first five rows
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [26]:
# Search for products in the same department using the loc function (locate a column) and isin (look within a lsit ie isin([17,18,19])) then view the first five rows
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## 08. Querying Data

1. If you haven't done so already, perform the wrangling procedures you walked through in this Exercise on your project data in a new notebook for this Exercise **see steps above**.Then, add a new section heading to separate your wrangling procedures from the procedures you’ll be conducting in **the steps below**.

2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [27]:
# Dataframe information
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(1)
memory usage: 156.6+ MB


In [28]:
# Change the order_id column and view the data types
df_ords['user_id'] = df_ords['user_id'].astype('str')
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(2)
memory usage: 156.6+ MB


In [29]:
# Check data type of only the order_id column
df_ords['user_id'].dtype

dtype('O')

3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [30]:
# Rename the order_dow column to orders_day_of_week without overwriting the original colum name (ie not with inplace = True) and view the first five rows of the renamed dataframe
df_ords_renamed = df_ords.rename(columns = {'order_hour_of_day' : 'hour_order_placed'}, inplace = False)
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [31]:
# View value counts for the order_hour_of_day column
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

* **The busiest hour of the day is between 10 and 11 am with 288418 orders placed.**

5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [32]:
# Look up department_id 4 using the data dictionary
print(data_dict_IC.get('4'))

{'department': 'produce'}


* **department_id 4 is produce**

6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [33]:
# Look up department breakfast using the data dictionary
data_dict_IC

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [34]:
# Search for breakfast products in the same department and view 
# (Alternative method df_breakfast =  df_prods[df_prods['department_id'].isin[14])
df_breakfast =  df_prods[df_prods['department_id']==14]
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

* **The breakfast dataframe has 1116 rows**

7. They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [35]:
# Search for products in the following departments (alcohol - 5, beverages - 7, meat seafood - 12 and deli - 20) using the loc function (locate a column) and isin (look within a lsit ie isin([17,18,19])) then view the first five rows
df_dinner_party_foods= df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]
df_dinner_party_foods

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [36]:
df_dinner_party_foods.shape

(7650, 5)

* **the dinner party foods dataframe has 7650 rows**<br> 
  [*Also provided above at the bottom of the dataframe (7650 rows x 5 columns)*]

9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.<br>
*(**NB** user_id "1." does not exist and I am assuming is is referring to user_id "1".)*

In [37]:
# Search for products in the same department using the loc function (locate a column) and isin (look within a lsit ie isin([17,18,19])) then view the first five rows
df_user_id_1 = df_ords.loc[df_ords['user_id']=='1']
df_user_id_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [38]:
#Basic statistics about user_id '1' order history
df_user_id_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder. <br>
13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format. <br>

In [39]:
# Export the df_ords dataframe into the Prepared_Data folder as orders_wrangled.csv
df_ords.to_pickle((os.path.join(path, '02_Data','02_2_Prepared_Data', 'orders_wrangled.pkl')))
# Export the df_dep_t_new dataframe into the Prepared_Data folder as departments_wrangled.csv
df_dep_t_new.to_pickle((os.path.join(path, '02_Data','02_2_Prepared_Data', 'departments_wrangled.pkl')))